jni/src/common/Foreign/JNI/Unsafe/Introspection.hs

```diff
@@ -0,0 +1,111 @@
+import Data.ByteString.Lazy (toStrict)
+import Data.Singletons
+import Data.Text (unpack)
+import Data.Text.Encoding (decodeUtf16LE)
+import Foreign.JNI.Unsafe.Internal
+import Foreign.JNI.Types
+import qualified Foreign.JNI.String as JNI
+import Foreign.Marshal.Array
+import Prelude hiding (String)
+
+-- | The "Class" class.
+kclass :: IO JClass
+kclass = findClass $ referenceTypeName $ sing @('Class "java.lang.Class")
+
+-- | The "Method" class
+kmethod :: IO JClass
+kmethod = findClass $ referenceTypeName $ sing @('Class "java.lang.reflect.Method")
+
+-- | Class.getMethods
+classGetMethodsMethod :: IO JMethodID
+classGetMethodsMethod = do
+  klass <- kclass
+  let sig = methodSignature [] (SArray $ sing @('Class "java.lang.reflect.Method"))
+  getMethodID klass (JNI.fromChars "getMethods") sig
+
+-- | Method.toString
+methodToStringMethod :: IO JMethodID
+methodToStringMethod = do
+  klass <- kmethod
+  let sig = methodSignature [] (sing @('Class "java.lang.String"))
+  getMethodID klass (JNI.fromChars "toString") sig
+
+-- | Method.getName
+methodGetNameMethod :: IO JMethodID
+methodGetNameMethod = do
+  klass <- kmethod
+  let sig = methodSignature [] (sing @('Class "java.lang.String"))
+  getMethodID klass (JNI.fromChars "getName") sig
+
+-- | Class.getName
+classGetNameMethod :: IO JMethodID
+classGetNameMethod = do
+  klass <- kclass
+  let sig = methodSignature [] (sing @('Class "java.lang.String"))
+  getMethodID klass (JNI.fromChars "getName") sig
+
+-- | Class.isInstance
+classIsInstanceMethod :: IO JMethodID
+classIsInstanceMethod = do
+  klass <- kclass
+  let sig = methodSignature [SomeSing $ sing @('Class "java.lang.Object")] (sing @('Prim "boolean"))
+  getMethodID klass (JNI.fromChars "isInstance") sig
+
+-- | @getSignatures c methodName@ yields the Java signatures of overloadings of
+-- methods called @methodName@ in class @c@.
+getSignatures :: JClass -> JNI.String -> IO [JNI.String]
+getSignatures c methodName = do
+  id <- classGetMethodsMethod
+  array :: JObjectArray <- unsafeCast <$> callObjectMethod c id []
+  l <- getArrayLength array
+  names <- forM [0 .. l - 1] $ \i -> do
+    ithObj :: JObject <- getObjectArrayElement array i
+    id <- methodGetNameMethod
+    jName :: JString <- unsafeCast <$> callObjectMethod ithObj id []
+    name <- toString jName
+    if name == methodName
+      then do
+        id <- methodToStringMethod
+        jMethodName :: JString <- unsafeCast <$> callObjectMethod ithObj id []
+        Just <$> toString jMethodName
+      else return Nothing
+  return $ catMaybes names
+
+-- | Turns a JString into a JNI String.
+-- Tricky, because Java encodes string in utf-16
+toString :: JString -> IO JNI.String
+toString obj = do
+  chars <- getStringChars obj
+  length <- fromIntegral <$> getStringLength obj
+  words <- peekArray length chars
+  return $ JNI.fromChars $ unpack $ decodeUtf16LE $ toStrict $ toLazyByteString $ fold $ word16LE <$> words
```

Comment:
How about going through the route package jvm uses:
```Haskell
instance Reify Text where
    reify jobj = do
        sz <- getStringLength jobj
        cs <- getStringChars jobj
        txt <- Text.fromPtr cs (fromIntegral sz)
        releaseStringChars jobj cs
        return txt
```

First go to Text, and then use `Text.encodeUtf8` followed by `Foreign.JNI.String.fromByteString`.

Is JNI.String the best representation to use in building the exception? Maybe we could just use `Text`.

`toStrict . toLazyByteString` above is likely the identity function.

Also, I'm not sure the jvm code is correct, but we would need to test that in another issue.

Reply:
`Text` could indeed be used instead of `JNI.String` in building the exception.
However, I don't think the code above is correct either: Java encodes its strings as utf-16, we may have encoding issues if we don't account for this

Reply:
I created #163 to revise that code.